### PR TITLE
feat(frontend): support idle in transaction session timeout

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Resolve https://github.com/risingwavelabs/risingwave/issues/13940 https://github.com/risingwavelabs/risingwave/issues/13885
- Introduce session variable `idle_in_transaction_session_timeout ` and its default value is 60000(ms) i.e. 1min.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by fuyufjh** [comment]
IMO the major motivation of #13940 is avoiding "pinned Hummock epoch to last forever", which requires proactively timing and killing the transaction once it timed out. Can this PR achieve that?

**Comment by chenzl25** [comment]
> proactively

I would like to improve it. Currently, this PR needs another statement to trigger the abortion of that session/transaction.

**Review by TennyZhuang — APPROVED** [review]

File: src/utils/pgwire/src/pg_protocol.rs [hunk 1]
```diff
@@ -550,6 +550,12 @@ where
         record_sql_in_span(&sql);
         let session = self.session.clone().unwrap();
 
+        if session.check_idle_in_transaction_timeout() {
+            self.process_terminate();
+            return Err(PsqlError::SimpleQueryError(
```
Comment (BugenZhao) [inline src/utils/pgwire/src/pg_protocol.rs:555]: What about adding a new error variant? I believe the handling logic can be quite similar to handling `Panic` here.

https://github.com/risingwavelabs/risingwave/blob/240416f43bae35c11d1d8a6d9e081fdd42fcc6bb/src/utils/pgwire/src/pg_protocol.rs#L350-L360

**Comment by BugenZhao** [comment]
> > proactively
> 
> I would like to improve it. Currently, this PR needs another statement to trigger the abortion of that session/transaction.

I assume we can start a background task to monitor the timeout when a transaction is started.

**Comment by fuyufjh** [comment]
> > > proactively
> > 
> > 
> > I would like to improve it. Currently, this PR needs another statement to trigger the abortion of that session/transaction.
> 
> I assume we can start a background task to monitor the timeout when a transaction is started.

Yeah, that's one approach. In my mind, we can set up a timer background task for each transaction. 

In either way, I am afraid this implementation can't be compatible.

File: src/frontend/src/session.rs [hunk 2]
```diff
@@ -325,6 +326,24 @@ impl FrontendEnv {
...
+                tokio::time::interval(core::time::Duration::from_secs(5));
+            check_idle_txn_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Skip);
+            check_idle_txn_interval.reset();
+            loop {
+                check_idle_txn_interval.tick().await;
+                sessions.read().values().for_each(|session| {
+                    let _ = session.check_idle_in_transaction_timeout();
+                })
+            }
+        });
```
Comment (chenzl25) [inline src/frontend/src/session.rs:344]: Add a background monitor to unpin snapshots proactively

Comment (fuyufjh) [reply src/frontend/src/session.rs:344]: A little bit brute-force 🤣

File: src/utils/pgwire/src/pg_server.rs [hunk 3]
```diff
@@ -132,6 +138,12 @@ impl ExecContextGuard {
     }
 }
 
+impl Drop for ExecContext {
+    fn drop(&mut self) {
+        *self.last_idle_instant.lock() = Some(Instant::now());
```
Comment (fuyufjh) [inline src/utils/pgwire/src/pg_server.rs:143]: As the `last_idle_instant` is only updated when execution completes, if a batch query runs for a long time e.g. scanning a large amount of data, will it be killed?

File: src/frontend/src/session.rs (hunk 2 above)
Comment (BugenZhao) [reply src/frontend/src/session.rs:344]: It's okay to spawn many tasks in tokio. So what about one background task for one session? 🥲

File: src/utils/pgwire/src/pg_server.rs (hunk 3 above)
Comment (chenzl25) [reply src/utils/pgwire/src/pg_server.rs:143]: > As the `last_idle_instant` is only updated when execution completes, if a batch query runs for a long time e.g. scanning a large amount of data, will it be killed? 

According to the postgresql docs, this variable doesn't care about how long a statement is running, which should be handled by parameters like statement timeout

Comment (chenzl25) [reply src/utils/pgwire/src/pg_server.rs:143]: Yes, we won't kill any query, but unpin the snapshot (either by background monitor or session termination triggered by a later statement). The session would be terminated when the later statement arrives.

Comment (fuyufjh) [reply src/utils/pgwire/src/pg_server.rs:143]: > but unpin the snapshot

By "killed" I mean any means that cause the query to end. When a snapshot is unpinned, it's possible that the iterator runs into errors and terminates, is it the cases?

**Comment by BugenZhao** [comment]
> Currently, this PR needs another statement to trigger the abortion of that session/transaction.

Oops, this seems to be the behavior of Postgres. 😂

File: src/utils/pgwire/src/pg_server.rs (hunk 3 above)
Comment (chenzl25) [reply src/utils/pgwire/src/pg_server.rs:143]: I just ensure there is no running SQL in the current session during calling `check_idle_in_transaction_timeout`, so we don't need to worry about that.

**Review by fuyufjh — APPROVED** [review]
LGTM

**Pull request merged**